<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>ðŸ“¦ You can make an R package too!</title>
    <meta charset="utf-8" />
    <meta name="author" content="Emma Rand Department of Biology, University of York, UK" />
    <script src="libs/header-attrs/header-attrs.js"></script>
    <link href="libs/panelset/panelset.css" rel="stylesheet" />
    <script src="libs/panelset/panelset.js"></script>
    <link rel="stylesheet" href="style-xaringanthemer.css" type="text/css" />
    <link rel="stylesheet" href="style-custom.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# ðŸ“¦ You can make an R package too!
## DDA: Data Science for Modern &amp; open Research
### Emma Rand<br>Department of Biology, University of York, UK
### <br>ðŸ”— <a href="https://bit.ly/3mma-dda">bit.ly/3mma-dda</a>

---




layout: true
  
&lt;div class="my-footer"&gt;
&lt;span&gt;
&lt;a href="https://bit.ly/3mma-dda" target="_blank"&gt;ðŸ”— bit.ly/3mma-dda&lt;/a&gt;
&lt;/span&gt;
&lt;/div&gt; 

&lt;!-- Presenting on zoom or similar? Press C to clone your slides, then P for presenter mode for one set. You can share the whole screen version of the slides while still seeing your presenter view --&gt;

---

## I want to encourage...


.pull-left[
.center[

.large[
&amp;nbsp;
]

&amp;nbsp;

.midi[

R users to make packages

]


.small[

why?  
when?  
how?  

]

]
]

--

.pull-right[
.center[
.large[
Use of scripting
]

People to use R

.midi[

R users to make packages

]


.small[

why?  
when?  
how?  

]

]

]
---
class: middle, inverse

# About you!

---

Go to https://www.menti.com/ and use the code 5900 8511

&lt;img src="images/mentimeter_qr_code.png" width="48%" style="display: block; margin: auto;" /&gt;

---

## You are not alone!

Expectations of Stage 1 Biologists about Data Analysis.

Responses to the question *"I will enjoy data analysis"*

--

&lt;img src="images/enjoy_2018.png" width="100%" style="display: block; margin: auto;" /&gt;

---

## You are not alone!


"*Nobody wants to learn to code. They just want to kick ass*."

&lt;img src="images/james-long.png" width="20%" style="display: block; margin: auto;" /&gt;

.small[
James Long rstudio::conf 2019
Putting empathy in action Building a 'community of practice' for analytics in a global corporation. 
]

---

## Me

.pull-left[

Biologist  
Data Analysis &amp; reproducibility  
Training early career researchers  
[Software Sustainability Institute](https://www.software.ac.uk/) Fellow  
.small[*Good software practices create better software, better software improves the reproducibility and reusability of research.*]
]

--

.pull-right[


[R Forwards](https://forwards.github.io/) 
.small[The R Foundation's task force for widening participation in the R community, especially contributers.]
&lt;img src="images/rforwards.png" width="15%" style="display: block; margin: auto;" /&gt;

[Cloud-SPAN](https://cloud-span.york.ac.uk/) 
.small[Trains researchers, and the research software engineers that support them, to run 'omics analyses on cloud-HPC.]
&lt;img src="images/transparent-cloud-only.PNG" width="20%" style="display: block; margin: auto;" /&gt;

]



---
class: middle, inverse

# Why script?


---
## Reproducibility is key!

One definition
*"... obtaining consistent results using the same input data; computational steps, methods, and code; and conditions of analysis."*. (National Academies of Sciences, Engineering Medicine, et al., 2019)

Also see 

-  The Turing Way "Definitions for Reproducibility" (Arnold Bowler, et al., 2019)
-  National Science Foundation (Bollen Cacioppo, et al., 2015)

---
## Reproducibility is key!

&lt;img src="images/reproducible-matrix.jpg" width="700px" style="display: block; margin: auto;" /&gt;

.small[
How the Turing Way (Arnold Bowler, et al., 2019) defines reproducible research
]


---

## Reproducibility is key!

* Many high profile cases of work which did not reproduce e.g. Anil Potti unravelled by Baggerly and Coombes (2009) 




* **Will** become standard in Science and publishing e.g OECD Global Science Forum Building digital workforce capacity and skills for data-intensive science (OECD Global Science Forum, 2020)

---

## It's science!

&lt;iframe src="https://giphy.com/embed/qCj1NK1rxtnna" width="480" height="270" frameBorder="0" class="giphy-embed" allowFullScreen&gt;&lt;/iframe&gt;&lt;p&gt;&lt;a href="https://giphy.com/gifs/transparent-qCj1NK1rxtnna"&gt;via GIPHY&lt;/a&gt;&lt;/p&gt;

---

## Be nice to future you!

&lt;img src="images/future_you.png" width="360px" style="display: block; margin: auto;" /&gt;


---

## Be nice to future you!


&lt;iframe src="https://giphy.com/embed/M11UVCRrc0LUk" width="480" height="277" frameBorder="0" class="giphy-embed" allowFullScreen&gt;&lt;/iframe&gt;&lt;p&gt;&lt;a href="https://giphy.com/gifs/M11UVCRrc0LUk"&gt;via GIPHY&lt;/a&gt;&lt;/p&gt;

--

Five selfish reasons to work reproducibly (Markowetz, 2015a). Alternatively, see the [talk](https://youtu.be/yVT07Sukv9Q)

---

## Rationale for scripting analysis

Science is the generation of ideas, designing work to test them and reporting the results.

&lt;img src="images/rationale1.png" width="600px" style="display: block; margin: auto;" /&gt;

.pull-left[
Generating the results
]

.pull-right[
Analysing and reporting them
]
---

## Rationale for scripting analysis

We ensure **replicability** of laboratory and field work by planning and recording in lab books and using standard protocols. 


&lt;img src="images/rationale2.png" width="600px" style="display: block; margin: auto;" /&gt;

Even so replicating results can be hard.
---

## Rationale for scripting analysis

We ensure **replicability** of laboratory and field work by planning and recording in lab books and using standard protocols.  

&lt;img src="images/rationale3.png" width="600px" style="display: block; margin: auto;" /&gt;


Workflows for data analysis and reporting can, and should, be 100% **reproducible**! Scripting is a way to achieve this.

---

## Rationale for scripting analysis

That reproducibility applies to all aspects of the data workflow.

&lt;img src="images/rationale5.png" width="800px" style="display: block; margin: auto;" /&gt;

--

From importing or collecting the data, processing it for analysis, building statistical models and communicating the methods and results.


---
class: middle, inverse

# Why R?


---

## Why R?

Open source and free

--

-  leverages all willing expertise

-  problems are solved quickly

-  innovations appear often and quickly

-  fosters collaboration and community

-  abundant support

---

## Why R?

Open source and free

--
 .......But so is Python
  
--

R has reputation for catering to users who do not see themselves as programmers, and allowing them to slide gradually into programming.


&lt;img src="images/biologist1.png" width="700px" style="display: block; margin: auto;" /&gt;


---

## Why R?

Domain specific

--

Designed for data analysis and graphics - which means it is often easier to achieve those tasks in R than a general purpose programming language.

---

## Why R?

The R community is one of R's greatest assets. Vibrant, supportive, inclusive.

.pull-left[
* [#rstats](https://twitter.com/hashtag/rstats?lang=en) on twitter is very active
* [RForwards](https://forwards.github.io/about/) 
* [RLadies](https://rladies.org/) gender diversity promotion 
* [Hey! You there! You are welcome here](https://ropensci.org/blog/2017/06/23/community/)  
.footnote[
Artwork by @allison_horst 
]
]

.pull-right[
&lt;img src="images/welcome_to_rstats_twitter.png" width="400px" style="display: block; margin: auto;" /&gt;

]

---
class: middle, inverse

# Why make a package?

---

## Why make a package?

Conventionally:

-  Package developers  

-  Generalisable analytical methods  

-  For use on other data  

-  Public release


---

## Why make a package?

Package up your data analysis project!

-  You don't need to have a collection of highly generalisation functions

-  You don't need to share it with anyone else

-  If you are already trying to work reproducibly you are almost doing it anyway!

-  But it will help you do it better


---

## Why make a package?

If you are already trying to work reproducibly you are almost doing it anyway!

```
/stem-cell-proteomic
   /data-raw
   /date-processed
   /figures
   /R
   README.md
   report.Rmd
   stem-cell-proteomic.RProj

```
â€¦ making a package is just a short step beyond that


---

## I don't do *any* coding...


..and I'm not about to start!

--

You can still make steps towards reproducibility

-  organise your data (Wickham, 2014; Broman and Woo, 2018)
-  describe your data with metadata
-  have a data management plan e.g. [UK Data Service Checklist](https://ukdataservice.ac.uk/learning-hub/research-data-management/plan-to-share/checklist/)
-  have a consistent directory structure
-  have a consistent naming system e.g see Jenny Bryan's [Naming things](http://www2.stat.duke.edu/~rcs46/lectures_2015/01-markdown-git/slides/naming-slides/naming-slides.pdf)
-  Have a look at [Reproducible Research in R](https://r-cubed.rostools.org/index.html)


---
class: middle, inverse

# How to make a package?


---

## How to make a package?

Easier than you think!

-  Use RStudio
-  Use **`devtools`** - many helper functions


---

## How to make a package?

Create a package!


```r
usethis::create_package("myanalysis")
```

---

## How to make a package?

.pull-left[

&lt;img src="images/pkg.png" width="400px" style="display: block; margin: auto;" /&gt;

]

.pull-right[

-  It is a package and an RStudio project.
-  DESCRIPTION provides metadata about your package.
-  R/ directory for code
-  NAMESPACE declares the functions you export and the external functions your work depends on

]


---

## How to make a package?

Metadata in DESCRIPTION

-  Title  
-  Version  
-  Authors  
-  Description  
-  License  


---

## How to make a package?

Document your package!


```r
devtools::document()
```


---

## How to make a package?

Check your package!


```r
devtools::check()
```

---

class: middle, inverse

# Summary

---

## Summary

-  Few biologists want to learn to code  
-  We do want our work to reproduce  
-  We can take steps towards reproducibility  
-  R is cool!
-  If you are an R user, you can make a package  
-  Encourages and supports reproducibility  
-  Helps you document your analysis  
-  Helps you test and check your analysis  
-  Is made easy with devtools and friends  

---

class: middle, inverse

# What questions do you have?

---

class: middle, inverse

# Acknowledgements

---

Slides made with: **`knitr`** (Xie, 2020; Xie, 2015; Xie, 2014),**`R Markdown`** (Allaire Xie, et al., 2019; Xie Allaire, et al., 2018), **`xaringan`** (Xie, 2019), **`xaringanthemer`** (Aden-Buie, 2020a), **`xaringanExtra`** (Aden-Buie, 2020c)

Referencing with **`RefManageR`** (McLean, 2017).

---

## References

.tiny[
Aden-Buie, G. (2020c). _xaringanExtra: Extras And
Extensions for Xaringan Slides_. R package version
0.2.3.9000. URL:
[https://github.com/gadenbuie/xaringanExtra](https://github.com/gadenbuie/xaringanExtra).

Aden-Buie, G. (2020a). _xaringanthemer: Custom 'xaringan'
CSS Themes_. R package version 0.3.0. URL:
[https://CRAN.R-project.org/package=xaringanthemer](https://CRAN.R-project.org/package=xaringanthemer).

Allaire, J., Y. Xie, et al. (2019). _rmarkdown: Dynamic
Documents for R_. R package version 1.16. URL:
[https://github.com/rstudio/rmarkdown](https://github.com/rstudio/rmarkdown).

Arnold, B., L. Bowler, et al. (2019). _The Turing Way: A
Handbook for Reproducible Data Science_. This work was
supported by The UKRI Strategic Priorities Fund under the
EPSRC Grant EP/T001569/1, particularly the "Tools,
Practices and Systems" theme within that grant, and by
The Alan Turing Institute under the EPSRC grant
EP/N510129/1.. Zenodo. DOI:
[10.5281/zenodo.3233986](https://doi.org/10.5281%2Fzenodo.3233986).
URL:
[https://doi.org/10.5281/zenodo.3233986](https://doi.org/10.5281/zenodo.3233986).

Baggerly, K. A. and K. R. Coombes (2009). "DERIVING
CHEMOSENSITIVITY FROM CELL LINES: FORENSIC BIOINFORMATICS
AND REPRODUCIBLE RESEARCH IN HIGH-THROUGHPUT BIOLOGY".
In: _Ann. Appl. Stat._ 3.4, pp. 1309-1334.

Bollen, K., J. T. Cacioppo, et al. (2015). _Social,
Behavioral, and Economic Sciences Perspectives on Robust
and Reliable Science_. National Science Foundation.

Broman, K. W. and K. H. Woo (2018). "Data Organization in
Spreadsheets". In: _Am. Stat._ 72.1, pp. 2-10.

Markowetz, F. (2015a). "Five selfish reasons to work
reproducibly". En. In: _Genome Biol._ 16, p. 274.

McLean, M. W. (2017). "RefManageR: Import and Manage
BibTeX and BibLaTeX References in R". In: _The Journal of
Open Source Software_. DOI:
[10.21105/joss.00338](https://doi.org/10.21105%2Fjoss.00338).
]

---

## References

.tiny[
National Academies of Sciences, Engineering, Medicine, et
al. (2019). _Understanding Reproducibility and
Replicability_. National Academies Press (US).

OECD Global Science Forum (2020). _Building digital
workforce capacity and skills for data-intensive
science_. OECD.

Wickham, H. (2014). "Tidy Data". In: _Journal of
Statistical Software_ 59.10, p. 1Ã¢â‚¬â€œ23. DOI:
[10.18637/jss.v059.i10](https://doi.org/10.18637%2Fjss.v059.i10).
URL:
[https://www.jstatsoft.org/index.php/jss/article/view/v059i10](https://www.jstatsoft.org/index.php/jss/article/view/v059i10).

Xie, Y. (2014). "knitr: A Comprehensive Tool for
Reproducible Research in R". In: _Implementing
Reproducible Computational Research_. Ed. by V. Stodden,
F. Leisch and R. D. Peng. ISBN 978-1466561595. Chapman
and Hall/CRC. URL:
[http://www.crcpress.com/product/isbn/9781466561595](http://www.crcpress.com/product/isbn/9781466561595).

Xie, Y. (2015). _Dynamic Documents with R and knitr_.
2nd. ISBN 978-1498716963. Boca Raton, Florida: Chapman
and Hall/CRC. URL:
[https://yihui.org/knitr/](https://yihui.org/knitr/).

Xie, Y. (2019). _xaringan: Presentation Ninja_. R package
version 0.12. URL:
[https://CRAN.R-project.org/package=xaringan](https://CRAN.R-project.org/package=xaringan).

Xie, Y. (2020). _knitr: A General-Purpose Package for
Dynamic Report Generation in R_. R package version 1.30.
URL:
[https://yihui.org/knitr/](https://yihui.org/knitr/).

Xie, Y., J. Allaire, et al. (2018). _R Markdown: The
Definitive Guide_. ISBN 9781138359338. Boca Raton,
Florida: Chapman and Hall/CRC. URL:
[https://bookdown.org/yihui/rmarkdown](https://bookdown.org/yihui/rmarkdown).
]

---

## License

Licensed under a &lt;a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/"&gt;Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License&lt;/a&gt;.
&lt;a rel="license" href="http://creativecommons.org/licenses/by-nc-sa/4.0/"&gt;&lt;img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-nc-sa/4.0/88x31.png" /&gt;&lt;/a&gt;&lt;br /&gt;
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"ratio": "16:9",
"highlightLines": true,
"highlightStyle": "github",
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
